extensions/wikia/AdEngine3/src/templates/fmr-rotator.js

```diff
@@ -0,0 +1,237 @@
+	const btRecNodeInViewport = btRec && currentRecNode && utils.isInViewport(currentRecNode);
+	const adSlotInViewport = currentAdSlot && currentAdSlot.getElement() && utils.isInViewport(currentAdSlot.getElement());
+
+	return recirculationElementInViewport || btRecNodeInViewport || adSlotInViewport;
+}
+
+/**
+ * Update module refresh information after rotation
+ *
+ * @returns {void}
+ */
+function updateAdRefreshInformation() {
+	refreshInfo.lastRefreshTime = new Date();
+	refreshInfo.refreshAdPosition = window.scrollY;
+}
+
+/**
+ * Checks if user reached certain scroll position to enable FMR logic
+ *
+ * @returns {boolean}
+ */
+function isStartPositionReached() {
+	return refreshInfo.startPosition <= window.scrollY;
+}
+
+/**
+ * Checks if user scrolled some distance before slot reload
+ *
+ * @returns {boolean}
+ */
+function isScrolledEnoughToRotate() {
+	return Math.abs(window.scrollY - refreshInfo.refreshAdPosition) > refreshInfo.heightScrolledThreshold;
+}
+
+/*********************************
+ * Watchers or exported functions
+ *********************************/
+
+/**
+ * Rotates floating medrec based on given conditions
+ *
+ * @returns {void}
+ */
+function rotateSlots() {
+	if (!isRefreshLimitAvailable()) {
+		scrollListener.removeCallback(rotatorListener);
+	} else if (enoughTimeSinceLastRefresh() && isScrolledEnoughToRotate() && isInViewport()) {
+		if (refreshInfo.adVisible) {
+			if (btRec) {
+				removeRecNode();
+			} else {
+				slotTweaker.hide(currentAdSlot);
+			}
+
+			updateAdRefreshInformation();
+			swapRecirculation(true);
+		} else {
+			scrollListener.removeCallback(rotatorListener);
+
+			context.push('state.adStack', { id: nextSlotName });
+
+			applyRec(slotStatusChanged);
+		}
+	}
+}
+
+/**
+ * Wait for conditions and inject first floating medrec
+ *
+ * @returns {void}
+ */
+function showSlotWhenPossible() {
+	if (enoughTimeSinceLastRefresh() && isStartPositionReached()) {
+		scrollListener.removeCallback(rotatorListener);
+
+		context.push('state.adStack', { id: nextSlotName });
+
+		applyRec(slotStatusChanged);
+	}
+}
+
+/**
+ * Callback executed on ad slot or BT rec placement load
+ *
+ * @param {string} slotName
+ * @param {string} slotStatus
+ * @returns {void}
+ */
+function slotStatusChanged(slotName = fmrPrefix, slotStatus = AdSlot.STATUS_SUCCESS) {
+	if (!universalAdPackage.isFanTakeoverLoaded() && slotName.substring(0, 16) === fmrPrefix) {
+		if (!btRec) {
+			currentAdSlot = slotService.get(nextSlotName);
+			nextSlotName = fmrPrefix + (currentAdSlot.getConfigProperty('repeat.index') + 1);
+		}
+
+		updateAdRefreshInformation();
+
+		if (slotStatus === AdSlot.STATUS_SUCCESS) {
+			swapRecirculation(false);
+		}
+
+		rotatorListener = scrollListener.addCallback(() => rotateSlots());
+	}
+}
+
+/**
+ * Enable floating medrec rotating logic
+ *
+ * @param {string} slotName
+ * @returns {void}
+ */
+export function rotateIncontentBoxad(slotName) {
```

Comment:
It isn't really template (whole file). Could we place it somewhere else?

Reply:
Moved.